File: Problems/0949_Largest_Time_for_Given_Digits/Lagest_Time_for_Given_Digits.py
# coding: utf-8

import os
import sys
import time
from itertools import permutations

class Solution:
    def largestTimeFromDigits(self, A):
        """
        :type A: List[int]
        :rtype: str
        """
        k = sorted(list(permutations(A)), reverse=True)
        for i in k:            
            a,b,c,d = i
            su = (a*10 + b)
            sd = (c*10 + d) 
            if su < 24 and sd <60:
                return  f"{a}{b}:{c}{d}"
        return ''

def str_to_int_array(flds):
    nums = [0]*len(flds)
    for i in range(len(flds)):
        nums[i] = int(flds[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace("[","").replace("]","").rstrip().split(",")
    A = str_to_int_array(flds)

    time0 = time.time()

    sl = Solution()
    result = sl.largestTimeFromDigits(A)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()

File: Problems/0900_0999/0949_Largest_Time_for_Given_Digits/Project_CS/Lagest_Time_for_Given_Digits.cs
using System;
using System.Linq;

public class Solution
{
    public string LargestTimeFromDigits(int[] A)
    {
        Array.Sort(A);
        Array.Reverse(A);
        int su = 0, sd = 0;
        for (int i1 = 0; i1 < A.Length; ++i1)
        {
            for (int i2 = 0; i2 < A.Length; i2++)
            {
                if (i2 == i1)
                    continue;
                for (int i3 = 0; i3 < A.Length; ++i3)
                {
                    if (i3 == i1 || i3 == i2)
                        continue;
                    for (int i4 = 0; i4 < A.Length; ++i4)
                    {
                        if (i4 == i1 || i4 == i2 || i4 == i3)
                            continue;
                        su = (A[i1]*10 + A[i2]);
                        sd = (A[i3]*10 + A[i4]);
                        if (su < 24 && sd < 60)
                            return A[i1].ToString() + A[i2].ToString() + ":" + A[i3].ToString() + A[i4].ToString();
                    }
                }
            }
        }
        return "";
    }

    public int[] str_to_int_array(string[] flds)
    {
        if (flds.Length <= 0)
            return null;

        int[] nums = new int[flds.Length];
        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] flds)
    {
        string results = "";
        if (flds.Length <= 0)
            return results;
        
        results = flds[0].ToString();
        for (int i = 1; i < flds.Length; ++i)
        {
            results += ", " + flds[i].ToString();
        }

        return results;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[","").Replace("]","").Trim().Split(',');

        int[] A = str_to_int_array(flds);
        Console.WriteLine("A = " + output_int_array(A));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        string results = LargestTimeFromDigits(A);
        Console.WriteLine("result = " + results);
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
